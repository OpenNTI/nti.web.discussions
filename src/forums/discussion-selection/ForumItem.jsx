import React from 'react';
import PropTypes from 'prop-types';

import HighlightedContent from './HighlightedContent';


export default class ForumItem extends React.Component {
	static propTypes = {
		item: PropTypes.shape({
			title: PropTypes.string,
			get: PropTypes.func.isRequired
		}).isRequired,
		searchTerm: PropTypes.string,
		onClick: PropTypes.func,
	}

	onClick = () => this.props.onClick(this.props.item);

	render () {
		const { item, searchTerm } = this.props;

		return (
			<div className="discussion-selection-item" onClick={this.onClick}>
				<div className="content">
					<HighlightedContent content={item.get('displayTitle') || item.title} term={searchTerm} />
				</div>
				<div className="arrow-icon">
					<i className="icon-chevron-right" />
				</div>
			</div>
		);
	}
}
