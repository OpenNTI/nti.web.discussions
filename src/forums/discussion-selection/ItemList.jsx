import React from 'react';
import PropTypes from 'prop-types';

import { filterItemsBySearchTerm } from './utils';
import { ContainerBase, Empty, Header } from './parts';
import Item from './Item';

const Container = styled(ContainerBase)`
	margin-top: 30px;
`;


List.propTypes =
ItemList.propTypes = {
	items: PropTypes.arrayOf(PropTypes.object),
	headerText: PropTypes.string,
	searchTerm: PropTypes.string,
	onSelect: PropTypes.func,
	ItemCmp: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
};

ItemList.defaultProps = {
	ItemCmp: Item
};

/**
 * @deprecated This code is retained for legacy content-backed discussions.
 */
export default function ItemList ({headerText: heading, ...props}) {

	return (
		<Container data-testid="discussion-selection-item-container">
			<Header>{heading}</Header>
			<List {...props}/>
		</Container>
	);
}

function List ({ items, onSelect, searchTerm, ItemCmp }) {

	const filteredItems = filterItemsBySearchTerm(items, searchTerm);

	if(!filteredItems?.length) {
		return <Empty>No discussions found</Empty>;
	}

	return filteredItems.map((item, i) => (
		<ItemCmp
			key={i}
			item={item}
			onClick={onSelect}
			searchTerm={searchTerm}
		/>
	)
	);
}
