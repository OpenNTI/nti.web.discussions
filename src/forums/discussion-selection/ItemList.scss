@import "~nti-style-common";

.discussion-selection-item {
	cursor: pointer;
	width: 246px;
	height: 50px;
	background: white;
	margin: 0 10px 10px 0;
	padding: 12px;
	position: relative;
	float: left;
	-webkit-box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.28);
	box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.28);

	.content {
		float: left;
		position: relative;
	}

	.arrow-icon {
		float: right;
		position: relative;
	}
}

.discussion-selection-item-container {
	margin-top: 30px;

	.header {
		margin-top: 10px;
		margin-bottom: 10px;
		font-size: 18px;
		font-weight: 600;
		color: #494949;
	}

	.no-results {
		position: relative;
		float: left;
	}
}
