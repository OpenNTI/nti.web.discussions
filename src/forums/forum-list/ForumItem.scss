@import '~@nti/style-common';

.forum-item-li {
	border-bottom: 1px dotted #d8d8d8;

	.forum-link {
		padding: .2rem 0 .2rem 0.9375rem;
		display: block;

		@include responsive-range(mobile) {
			padding: .625rem 0 .625rem 0.9375rem;
		}

		&.active {
			background: #3fb3f6;
			box-shadow: inset 3px 0 0 0 #107eed;
			.item-container {
				.item-main {
					.title {
						color: white;
					}

					.meta {
						color: white;
					}
				}

				.forum-item-edit {
					color: white;
				}
			}
		}
	}

	.item-container {
		display: flex;
		flex-direction: row;

		.item-main {
			flex-grow: 1;

			.title {
				display: block;
				font-size: rem-calc(14px);
				color: var(--primary-grey);
				font-family: "Open Sans Condensed";
				font-weight: 600;
				overflow: hidden;
				text-overflow: ellipsis;
				max-width: 200px;
				white-space: nowrap;

				@include responsive-range(mobile) {
					font-size: 1rem;
				}
			}

			.meta {
				font-size: .75rem;
				display: block;
				color: $tertiary-grey;
				font-family: 'Open Sans';
				font-weight: 600;

				@include responsive-range(mobile) {
					padding: .5rem 0;
				}
			}
		}

		.forum-item-reports {
			display: none;
			justify-content: center;
			align-items: center;
			padding: 0 rem-calc(5px);
			color: #ddd;
		}

		&:hover {
			.forum-item-reports {
				display: flex;
			}
		}
	}
}
