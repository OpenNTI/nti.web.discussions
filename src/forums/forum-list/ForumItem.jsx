import './ForumItem.scss';
import React from 'react';
import PropTypes from 'prop-types';
import {decorate} from '@nti/lib-commons';
import { LinkTo } from '@nti/web-routing';
import { HOC } from '@nti/web-commons';
import { scoped } from '@nti/lib-locale';
import { Viewer } from '@nti/web-reports';

const DEFAULT_TEXT = {
	count: {
		zero: 'No Discussions',
		one: '1 Discussion',
		other: '%(count)s Discussions'
	}
};

const t = scoped('forums.topic', DEFAULT_TEXT);

class ForumItem extends React.Component {
	static propTypes = {
		item: PropTypes.shape({
			displayTitle: PropTypes.string,
			title: PropTypes.string.isRequired,
			edit: PropTypes.func.isRequired,
			TopicCount: PropTypes.number.isRequired,
			hasLink: PropTypes.func.isRequired,
			Links: PropTypes.array.isRequired,
			Reports: PropTypes.array
		}).isRequired,
	}

	static contextTypes = {
		router: PropTypes.object
	}

	showReports = (e) => {
		e.preventDefault();

		const report = this.props.item.Reports[0];

		Viewer.show(report);
	}

	render () {
		const { item } = this.props;

		const canShowReports = item.Links.some(x => x.rel.startsWith('report-')) && item.Reports;

		return (
			<li className="forum-item-li">
				<LinkTo.Object object={item} activeClassName="active" className="forum-link">
					<div className="item-container">
						<div className="item-main">
							<span className="title">{item.displayTitle}</span>
							<div className="meta">
								<span className="see-all count">{t('count', { count: item.TopicCount })}</span>
							</div>
						</div>
						{canShowReports && (
							<div className="forum-item-reports" onClick={this.showReports}>
								<i className="icon-report" />
							</div>
						)}
					</div>
				</LinkTo.Object>
			</li>
		);
	}
}


export default decorate(ForumItem, [
	HOC.ItemChanges.compose
]);
