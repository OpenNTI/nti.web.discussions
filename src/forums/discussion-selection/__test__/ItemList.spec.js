/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import ItemList from '../ItemList';
import ForumItem from '../ForumItem';

describe('Item list test', () => {
	test('Simple list', () => {
		const items = [
			{ title: 'item 1' },
			{ title: 'item 2' },
			{ title: 'item 3' }
		];

		const {container} = render(<ItemList items={items} searchTerm="item"/>);
		expect(container.querySelectorAll('.discussion-selection-item').length).toBe(3);

		const tree = renderer.create(<ItemList items={items} searchTerm="item" />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	test('Filtered', () => {
		const items = [
			{ title: 'item 1' },
			{ title: 'item 2' },
			{ title: 'item 3' }
		];

		const {container} = render(<ItemList items={items} searchTerm="2"/>);
		expect(container.querySelectorAll('.discussion-selection-item').length).toBe(1);

		const tree = renderer.create(<ItemList items={items} searchTerm="2" />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	test('NTI-6655: Open Bin Forums', () => {
		const items = [
			{
				MimeType: 'application/vnd.nextthought.forums.communityforum',
				NTIID: 'tag:nextthought.com,2011-10:unknown-OID-0x055933:5573657273:hC4MEZYP0nC',
				title: 'Open Announcements',
				displayTitle: 'Annoucments',
				ContainerId: 'tag:nextthought.com,2011-10:Public-OID-0x05592f:5573657273:hC4MEZYP0nF',
				get: function (property) { return this[property]; }
			},
			{
				MimeType: 'application/vnd.nextthought.forums.communityforum',
				NTIID: 'tag:nextthought.com,2011-10:unknown-OID-0x055930:5573657273:hC4MEZYP0nE',
				title: 'Open Discussions',
				displayTitle: 'Discussions',
				ContainerId: 'tag:nextthought.com,2011-10:Public-OID-0x05592f:5573657273:hC4MEZYP0nF',
				get: function (property) { return this[property]; }
			}
		];
		const tree = renderer.create(<ItemList items={items} ItemCmp={ForumItem} onSelect={() => {}} searchTerm="" />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
