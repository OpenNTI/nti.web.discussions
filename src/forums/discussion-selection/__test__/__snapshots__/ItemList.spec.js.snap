// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item list test Filtered 1`] = `
<div
  className="discussion-selection-item-container"
>
  <div
    className="discussion-selection-item"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div>
        <span>
          item 
        </span>
        <span
          className="discussion-selection-highlight"
        >
          2
        </span>
      </div>
    </div>
    <div
      className="arrow-icon"
    >
      <i
        className="icon-chevron-right"
      />
    </div>
  </div>
</div>
`;

exports[`Item list test NTI-6655: Open Bin Forums 1`] = `
<div
  className="discussion-selection-item-container"
>
  <div
    className="discussion-selection-item"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div>
        Annoucments
      </div>
    </div>
    <div
      className="arrow-icon"
    >
      <i
        className="icon-chevron-right"
      />
    </div>
  </div>
  <div
    className="discussion-selection-item"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div>
        Discussions
      </div>
    </div>
    <div
      className="arrow-icon"
    >
      <i
        className="icon-chevron-right"
      />
    </div>
  </div>
</div>
`;

exports[`Item list test Simple list 1`] = `
<div
  className="discussion-selection-item-container"
>
  <div
    className="discussion-selection-item"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div>
        <span
          className="discussion-selection-highlight"
        >
          item
        </span>
        <span>
           1
        </span>
      </div>
    </div>
    <div
      className="arrow-icon"
    >
      <i
        className="icon-chevron-right"
      />
    </div>
  </div>
  <div
    className="discussion-selection-item"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div>
        <span
          className="discussion-selection-highlight"
        >
          item
        </span>
        <span>
           2
        </span>
      </div>
    </div>
    <div
      className="arrow-icon"
    >
      <i
        className="icon-chevron-right"
      />
    </div>
  </div>
  <div
    className="discussion-selection-item"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div>
        <span
          className="discussion-selection-highlight"
        >
          item
        </span>
        <span>
           3
        </span>
      </div>
    </div>
    <div
      className="arrow-icon"
    >
      <i
        className="icon-chevron-right"
      />
    </div>
  </div>
</div>
`;
